DeleteUserDataActivity/utils/documentdb.ts

```diff
@@ -0,0 +1,109 @@
+import * as DocumentDb from "documentdb";
+import { Either, left, right } from "fp-ts/lib/Either";
+import { Option } from "fp-ts/lib/Option";
+import * as DocumentDbUtilsBase from "io-functions-commons/dist/src/utils/documentdb";
+
+export * from "io-functions-commons/dist/src/utils/documentdb";
+
+export const deleteDocument = (
+  client: DocumentDb.DocumentClient,
+  collectionUri: DocumentDbUtilsBase.IDocumentDbCollectionUri,
+  documentId: string,
+  partitionKey?: string
+): Promise<Either<DocumentDb.QueryError, string>> => {
+  const documentUri = DocumentDbUtilsBase.getDocumentUri(
+    collectionUri,
+    documentId
+  );
+  return new Promise(resolve =>
+    client.deleteDocument(documentUri.uri, { partitionKey }, err =>
+      resolve(err ? left(err) : right(documentId))
+    )
+  );
+};
+
+/**
+ *  Find all versions of a document.
+ *
+ *  Pass the partitionKey field / values if it differs from the modelId
+ *  to avoid multi-partition queries.
+ */
+export function findAllVersionsByModelId<T>(
+  client: DocumentDb.DocumentClient,
+  collectionUri: DocumentDbUtilsBase.IDocumentDbCollectionUri,
+  modelIdField: string,
+  modelIdValue: string,
+  partitionKeyField: string,
+  partitionKeyValue: string
+): DocumentDbUtilsBase.IResultIterator<T & DocumentDb.RetrievedDocument> {
+  return DocumentDbUtilsBase.queryDocuments(
+    client,
+    collectionUri,
+    {
+      parameters: [
+        {
+          name: "@modelId",
+          value: modelIdValue
+        },
+        {
+          name: "@partitionKey",
+          value: partitionKeyValue
+        }
+      ],
+      // do not use ${collectionName} here as it may contain special characters
+      query: `SELECT * FROM m WHERE (m.${modelIdField} = @modelId
+          AND m.${partitionKeyField} = @partitionKey) ORDER BY m.version DESC`
+    },
+    partitionKeyValue
+  );
+}
+
+export function deleteAllDocuments<T extends DocumentDb.RetrievedDocument>(
+  client: DocumentDb.DocumentClient,
+  collectionUri: DocumentDbUtilsBase.IDocumentDbCollectionUri,
+  documentIterator: DocumentDbUtilsBase.IResultIterator<T>
+): DocumentDbUtilsBase.IFoldableResultIterator<
+  Promise<ReadonlyArray<Either<DocumentDb.QueryError, string>>>
+> {
+  return DocumentDbUtilsBase.reduceResultIterator(
+    documentIterator,
+    (
+      prev: Promise<ReadonlyArray<Either<DocumentDb.QueryError, string>>>,
+      curr: T
+    ) =>
+      Promise.all([prev, deleteDocument(client, collectionUri, curr.id)]).then(
```

Comment:
```suggestion
      prev.then(() => deleteDocument(client, collectionUri, curr.id)).then(
```

Reply:
maybe I'm wrong (I'm taking a quick look) but if you do it like that you'll trigger two promises in parallel for each run, the next iteration you will call Promise.all() on the same promises and son on...

Reply:
The difference is: in the first case operations are executed independently, but results are collected sequentially; in the latter operations are executed sequentially.